### ISSUE TITLE
Enable `--asan` when compile macOS unit tests on CI

### ISSUE THREAD

**Issue by dkwingsmt** [flutter/flutter#142159]
Recently I've discovered that some code crashes when `--asan` is enabled. Our CI tests are not run with `--asan`, therefore these bugs were not caught. We should enable `--asan` when compiling macOS unit tests on CI.

Related: https://github.com/flutter/engine/pull/49995

**Comment by jmagman** [comment]
Some more info in https://github.com/flutter/flutter/issues/108469 and https://github.com/flutter/flutter/issues/78433

**Comment by dkwingsmt** [comment]
> Some more info in #108469 and #78433

Yeah, I'm aware and I didn't turn on asan for iOS either. But are we able to turn it on for macOS/unittests?

**Comment by cbracken** [comment]
I think the question is going to be whether the clang toolchain we get from fuchsia has macOS asan support enabled/working nicely. 

I've enabled asan on personal projects on macOS but as a caveat, I've only tried this with the toolchain vended from Xcode, not the one we use for Flutter.

**Comment by jmagman** [comment]
> Yeah, I'm aware and I didn't turn on asan for iOS either. But are we able to turn it on for macOS/unittests?

I guess I should have linked to https://github.com/flutter/flutter/issues/108469#issuecomment-1256697839 more specifically  which suggests using the dylib vended by Xcode, so for Xcode 15 and macOS would be:
`Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib`
